source/FeatureRegistry.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "FeatureRegistry.h"
+#include "logging/LoggerFactory.h"
+
+using namespace std;
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (features.count(name))
+    {
+        return features.at(name);
+    }
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::string name, std::shared_ptr<Feature> featureToAdd)
```

Comment:
It looks like we would need a "set" function to handle the cases where you enable a previously disabled feature. add() is only able to function upon startup since the features map hasn't been populated yet. We should discuss with Marco about whether we should create a new function or modify this one.

Reply:
I think add serves a good purpose of literally adding features to the registry. Maybe a new method set() or something similar will be useful for dynamic reload. However, I will add this in the dynamic reload PR because it serves no purpose as of now.